src/osd/OSD.cc

```diff
@@ -9233,10 +9233,12 @@ void OSDService::adjust_pg_priorities(vector<PG*> pgs, int newflags)
       // make sure the PG is in correct state before forcing backfill or recovery, or
       // else we'll make PG keeping FORCE_* flag forever, requiring osds restart
       // or forcing somehow recovery/backfill.
+      i->lock();
```

Comment:
As @jdurgin wrote, OFR_CANCEL needs similar lock  (and **_**change_recovery_force_mode).